torch/distributed/algorithms/ddp_comm_hooks/optimizer_overlap_hooks.py

```diff
@@ -40,23 +49,55 @@ def _apply_optim_in_backward_hook(
     optimizer with backward pass, DDP will run the below hook to run optimizer
     step for parameters after gradient communication has taken place.
     """
+    optim_in_bwd_state = OptimInBackwardHookState(
+        optim_stream=torch.cuda.Stream(),
+        wait_for_optim_stream_enqueued=False,
+    )
+
     def apply_optim_in_backward_hook(
-        hook_state: Any, bucket: dist.GradBucket,
+        hook_state: Any, bucket: dist.GradBucket, optim_stream_state,
     ) -> torch.futures.Future[torch.Tensor]:
         # Run original hook
         fut = hook(hook_state, bucket)
-        def optimizer_step(fut):
+        optimizer_stream = optim_stream_state.optim_stream
+        with torch.cuda.stream(optimizer_stream):
+            fut.wait()
             model_params = bucket.parameters()
             for p in model_params:
                 if hasattr(p, '_in_backward_optimizers'):
                     for optim in p._in_backward_optimizers:
                         optim.step()
 
-            return bucket.buffer()
+        # Need to return a Future[Tensor] to obey comm hook API contract.
+        fut = torch.futures.Future()
+        fut.set_result(bucket.buffer())
 
-        return fut.then(optimizer_step)
+        # enqueue a callback to wait for this optimizer stream at the end of
+        # backward.
+        def wait_for_optim_stream_callback():
+            torch.cuda.current_stream().wait_stream(
+                optim_stream_state.optim_stream
+            )
+            # reset for the next backwards pass
+            optim_stream_state.wait_for_optim_stream_enqueued = False
+
+        if not optim_stream_state.wait_for_optim_stream_enqueued:
+            Variable._execution_engine.queue_callback(
+                wait_for_optim_stream_callback
+            )
+            # mark that the callback is enqueued
+            optim_stream_state.wait_for_optim_stream_enqueued = True
+
+        return fut
```

Comment:
we do not need to queue call back, instead we can simply call " torch.cuda.current_stream().wait_stream(
                optim_stream_state.optim_stream)" in the final_backward call back?

Reply:
Is this referring to calling this in `finalize_backward` in C++? I agree we can do this, I am just not sure if we can pass a cuda stream from python to C++, but it should be possible.